File: monads-blockchain-explained/src/examples/ts/toy-monad.ts
class Monad<T> {
    private value: T;

    constructor(value: T) {
        this.value = value;
    }

    static of<U>(value: U): Monad<U> {
        return new Monad(value);
    }

    map<U>(fn: (value: T) => U): Monad<U> {
        return Monad.of(fn(this.value));
    }

    flatMap<U>(fn: (value: T) => Monad<U>): Monad<U> {
        return fn(this.value);
    }

    getValue(): T {
        return this.value;
    }
}

// Example usage:
const monad = Monad.of(5)
    .map(x => x + 2)
    .flatMap(x => Monad.of(x * 3));

console.log(monad.getValue()); // Output: 21

File: monads-blockchain-explained/src/examples/hs/toyMonad.hs
module ToyMonad where

-- A simple implementation of a toy monad in Haskell

-- The Monad type class
class Functor m => Monad m where
    return :: a -> m a
    (>>=) :: m a -> (a -> m b) -> m b

-- A simple Maybe-like monad
data Toy a = Nothing | Just a deriving (Show)

instance Functor Toy where
    fmap _ Nothing = Nothing
    fmap f (Just x) = Just (f x)

instance Monad Toy where
    return = Just
    Nothing >>= _ = Nothing
    (Just x) >>= f = f x

-- Example usage of the Toy monad
exampleUsage :: Toy Int
exampleUsage = do
    x <- return 5
    y <- return 10
    return (x + y)